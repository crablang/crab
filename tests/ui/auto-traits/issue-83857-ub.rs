#![allow(suspicious_auto_trait_impls)]

struct Always<T, U>(T, U);
unsafe impl<T, U> Send for Always<T, U> {}
struct Foo<T, U>(Always<T, U>);

trait False {}
unsafe impl<U: False> Send for Foo<u32, U> {}

trait WithAssoc {
    type Output;
}
impl<T: Send> WithAssoc for T {
    type Output = Self;
}
impl WithAssoc for Foo<u32, ()> {
    type Output = Box<i32>;
}

fn generic<T, U>(v: Foo<T, U>, f: fn(<Foo<T, U> as WithAssoc>::Output) -> i32) {
    //~^ ERROR `Foo<T, U>` cannot be sent between threads safely
    f(foo(v));
}

fn foo<T: Send>(x: T) -> <T as WithAssoc>::Output {
    x
}

fn main() {
    generic(Foo(Always(0, ())), |b| *b);
}
