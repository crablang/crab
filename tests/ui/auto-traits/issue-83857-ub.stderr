error[E0277]: `Foo<T, U>` cannot be sent between threads safely
  --> $DIR/issue-83857-ub.rs:20:38
   |
LL | fn generic<T, U>(v: Foo<T, U>, f: fn(<Foo<T, U> as WithAssoc>::Output) -> i32) {
   |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Foo<T, U>` cannot be sent between threads safely
   |
   = help: the trait `Send` is not implemented for `Foo<T, U>`
note: required for `Foo<T, U>` to implement `WithAssoc`
  --> $DIR/issue-83857-ub.rs:13:15
   |
LL | impl<T: Send> WithAssoc for T {
   |         ----  ^^^^^^^^^     ^
   |         |
   |         unsatisfied trait bound introduced here
help: consider introducing a `where` clause, but there might be an alternative better way to express this requirement
   |
LL | fn generic<T, U>(v: Foo<T, U>, f: fn(<Foo<T, U> as WithAssoc>::Output) -> i32) where Foo<T, U>: Send {
   |                                                                                +++++++++++++++++++++

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
